In [ ]:
import torch
torch.cuda.is_available()

!pip install -r /content/Hypertension-AI/requirements.txt

#!pip install -q -U transformers datasets accelerate peft trl bitsandbytes scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 MB 21.2 MB/s eta 0:00:00


In [ ]:
!python /content/Hypertension-AI/src/prepare_hypertension_llm_data.py \
  --data-path /content/Hypertension-AI/data/hypertension_dataset.csv \
  --output-dir /content/Hypertension-AI/data/llm \
  --test-size 0.2 \
  --valid-size 0.2 \
  --seed 42

{
  "train": 1270,
  "valid": 318,
  "test": 397,
  "stress_thresholds": {
    "low_max": 6.0,
    "moderate_max": 10.0
  }
}


In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm.py \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --train-file /content/Hypertension-AI/data/llm/train.jsonl \
  --valid-file /content/Hypertension-AI/data/llm/valid.jsonl \
  --output-dir /content/Hypertension-AI/artifacts/llm \
  --num-epochs 3 \
  --batch-size 2 \
  --gradient-accumulation 8 \
  --learning-rate 2e-4 \
  --max-length 512 \
  --max-new-tokens 16 \
  --eval-steps 100 \
  --logging-steps 20 \
  --save-steps 500 \
  --lora-r 16 \
  --lora-alpha 32 \
  --lora-dropout 0.05 \
  --load-in-4bit

2025-10-08 06:34:28.244378: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759905268.264130     772 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759905268.270169     772 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759905268.285640     772 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759905268.285663     772 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759905268.285667     772 computation_placer.cc:177] computation placer alr

{
  "adapter_dir": "/content/Hypertension-AI/artifacts/llm/lora_adapter",
  "validation_metrics": {
    "num_samples": 318,
    "hypertension_accuracy": 0.0,
    "stress_accuracy": 0.0,
    "joint_exact_match": 0.0
  }
}

In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm.py \
  --eval-only \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --max-new-tokens 16

2025-10-08 06:47:56.721690: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759906076.755947    4247 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759906076.766458    4247 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759906076.791405    4247 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906076.791435    4247 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906076.791443    4247 computation_placer.cc:177] computation placer alr

In [ ]:
!python /content/Hypertension-AI/src/evaluate_llm_detailed.py \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --data-file /content/Hypertension-AI/data/llm/test.jsonl \
  --max-new-tokens 24

2025-10-08 06:50:47.146170: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759906247.176412    5015 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759906247.182284    5015 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759906247.197724    5015 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906247.197750    5015 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906247.197755    5015 computation_placer.cc:177] computation placer alr



---


**Using the second version of Scripts**





In [ ]:
!python /content/Hypertension-AI/src/prepare_hypertension_llm_data_2.py \
  --data-path /content/Hypertension-AI/data/hypertension_dataset.csv \
  --output-dir /content/Hypertension-AI/data/llm \
  --base-model meta-llama/Llama-3.2-1B-Instruct

{
  "train": 1607,
  "valid": 179,
  "test": 199,
  "stress_thresholds": {
    "low_max": 6.0,
    "moderate_max": 10.0
  },
  "prompt_style": "chat_template_llama"
}


In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --train-file /content/Hypertension-AI/data/llm/train.jsonl \
  --valid-file /content/Hypertension-AI/data/llm/valid.jsonl \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --output-dir /content/Hypertension-AI/artifacts/llm \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --num-epochs 3 \
  --batch-size 2 \
  --gradient-accumulation 8 \
  --learning-rate 1e-4 \
  --max-length 512 \
  --max-new-tokens 24 \
  --load-in-4bit

2025-10-08 06:53:45.663254: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759906425.708301    5846 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759906425.718643    5846 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759906425.743412    5846 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906425.743450    5846 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759906425.743457    5846 computation_placer.cc:177] computation placer alr

In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --eval-only \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --max-new-tokens 24

2025-10-08 07:19:54.165555: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759907994.185738   12326 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759907994.191987   12326 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759907994.208469   12326 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759907994.208504   12326 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759907994.208508   12326 computation_placer.cc:177] computation placer alr

In [ ]:
!python /content/Hypertension-AI/src/evaluate_llm_detailed.py \
  --base-model meta-llama/Llama-3.2-1B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --data-file /content/Hypertension-AI/data/llm/test.jsonl \
  --max-new-tokens 24

2025-10-08 07:22:02.779724: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1759908122.799897   12879 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1759908122.805829   12879 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1759908122.820978   12879 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759908122.821005   12879 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1759908122.821009   12879 computation_placer.cc:177] computation placer alr



---



---

# Using 8B Model

In [ ]:
!python /content/Hypertension-AI/src/prepare_hypertension_llm_data_2.py \
  --data-path /content/Hypertension-AI/data/hypertension_dataset.csv \
  --output-dir /content/Hypertension-AI/data/llm \
  --base-model meta-llama/Llama-3.1-8B-Instruct

tokenizer_config.json: 100% 55.4k/55.4k [00:00<00:00, 88.3MB/s]
tokenizer.json: 100% 9.09M/9.09M [00:00<00:00, 12.9MB/s]
special_tokens_map.json: 100% 296/296 [00:00<00:00, 2.75MB/s]
{
  "train": 1607,
  "valid": 179,
  "test": 199,
  "stress_thresholds": {
    "low_max": 6.0,
    "moderate_max": 10.0
  },
  "prompt_style": "chat_template_llama"
}


In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --train-file /content/Hypertension-AI/data/llm/train.jsonl \
  --valid-file /content/Hypertension-AI/data/llm/valid.jsonl \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --output-dir /content/Hypertension-AI/artifacts/llm \
  --base-model meta-llama/Llama-3.1-8B-Instruct \
  --num-epochs 3 \
  --batch-size 2 \
  --gradient-accumulation 8 \
  --learning-rate 1e-4 \
  --max-length 512 \
  --max-new-tokens 24 \
  --load-in-4bit

2025-10-10 18:37:44.000316: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1760121464.021449    7429 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1760121464.027374    7429 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1760121464.042483    7429 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1760121464.042506    7429 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1760121464.042510    7429 computation_placer.cc:177] computation placer alr

In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --eval-only \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --base-model meta-llama/Llama-3.1-8B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --max-new-tokens 24

In [ ]:
!python /content/Hypertension-AI/src/evaluate_llm_detailed.py \
  --base-model meta-llama/Llama-3.1-8B-Instruct \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --data-file /content/Hypertension-AI/data/llm/test.jsonl \
  --max-new-tokens 24



---



---

# Using tensorblock/medllama3-v20-GGUF

In [ ]:
!python /content/Hypertension-AI/src/prepare_hypertension_llm_data_2.py \
  --data-path /content/Hypertension-AI/data/hypertension_dataset.csv \
  --output-dir /content/Hypertension-AI/data/llm \
  --base-model tensorblock/medllama3-v20-GGUF

You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama.LlamaTokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565 - if you loaded a llama tokenizer from a GGUF file you can ignore this message
Traceback (most recent call last):
  File "/content/Hypertension-AI/src/prepare_hypertension_llm_data_2.py", line 128, in <module>
    main()
  File "/content/Hypertension-AI/src/prepare_hypertension_llm_data_2.py", line 99, in main
    tokenizer = AutoTokenizer.from_pretrained(args.base_model, use_fast=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/transformers/

In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --train-file /content/Hypertension-AI/data/llm/train.jsonl \
  --valid-file /content/Hypertension-AI/data/llm/valid.jsonl \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --output-dir /content/Hypertension-AI/artifacts/llm \
  --base-model tensorblock/medllama3-v20-GGUF \
  --num-epochs 3 \
  --batch-size 2 \
  --gradient-accumulation 8 \
  --learning-rate 1e-4 \
  --max-length 512 \
  --max-new-tokens 24 \
  --load-in-4bit

2025-10-10 18:37:44.000316: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1760121464.021449    7429 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1760121464.027374    7429 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1760121464.042483    7429 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1760121464.042506    7429 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1760121464.042510    7429 computation_placer.cc:177] computation placer alr

In [ ]:
!python /content/Hypertension-AI/src/fine_tune_hypertension_llm_2.py \
  --eval-only \
  --test-file /content/Hypertension-AI/data/llm/test.jsonl \
  --base-model tensorblock/medllama3-v20-GGUF \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --max-new-tokens 24

In [ ]:
!python /content/Hypertension-AI/src/evaluate_llm_detailed.py \
  --base-model tensorblock/medllama3-v20-GGUF \
  --adapter-dir /content/Hypertension-AI/artifacts/llm/lora_adapter \
  --data-file /content/Hypertension-AI/data/llm/test.jsonl \
  --max-new-tokens 24